README
pk-update-icon
==============

Description
-----------

pk-update-icon displays notifications and an icon in the tray area of the panel
when package updates are available.

Build Instructions
------------------

pk-update-icon requires a POSIX:2004 compatible operating system, it has been
tested to work on Linux distributions.  The following tools and shared
libraries are required to build pk-update-icon:

- GNU make >= 3.81
- pkg-config
- GNU gettext
- GNU or BSD install
- GLib version 2.30 or later
- GTK+ version 2.24 or later
- libunique
- libnotify
- PackageKit-glib2

Rebuilding the man pages additionally requires the xsltproc tool from libxml2.

Before building pk-update-icon check the commented macros in the Makefile for
any macros you may need to override depending on the used toolchain and
operating system.

By default, all files will be installed under the "/usr/local" directory, a
different installation path prefix can be set via the `prefix` macro.  In
addition, a second path prefix can be specified via the `DESTDIR` macro which
will be prepended to any path, incuding the `prefix` macro path prefix.  In
contrast to `prefix`, the path specified via the `DESTDIR` macro will only be
prepended to paths during installation and not be used for constructing
internal paths.

The following instructions assume that `make` is GNU make, on some platforms
it may be installed under a different name or a non-default path.  In order to
start the build process run `make all`.  After a successful build, run `make
install` to install the program, any associated data files and the
documentation.

Previously built binaries, object files, generated data files and
documentation can be removed by running `make clean`, any additional,
generated files which are not removed by the `clean` target can be removed by
running `make clobber`.

Contact
-------

Please send any feedback, translations or bug reports via email to
<gber@opensuse.org>.

Bug Reports
-----------

When sending bug reports, please always mention the exact version of
pk-update-icon with which the issue occurs as well as the version of the
operating system you are using and make sure that you provide sufficient
information to reproduce the issue and include any input, output, any error
messages.

In case of build issues, please also specify the implementations and versions
of the tools and shared libraries used to build the program, in particular the
compiler.

In case of crashes, please generate a stack trace with a suitable debugger
such as gdb, lldb, dbx, or debug after a crash has occurred either by
examining the resulting core file or by running the program from the debugger
and attach it to the bug report.  In order to generate a meaningful stack
trace the program as well as any dynamically linked libraries need to be built
with debugging information, see the documentation of the used compiler for the
required compiler flags.  If any of the dynamically linked shared libraries do
not contain debugging information, please either install debugging information
for these libraries using mechanisms provided by your operating system or
rebuild the libraries accordingly.  Please refer to the documentation of the
debugger for detailed instructions on generating backtraces.

License
-------

Except otherwise noted, all files are Copyright (C) 2014 Guido Berhoerster and
distributed under the following license terms:

Copyright (C) 2014 Guido Berhoerster <gber@opensuse.org>

Licensed under the GNU General Public License Version 2

This program is free software; you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation; either version 2 of the License, or
(at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program; if not, write to the Free Software
Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.


NEWS
News
====

pk-update-icon 2 (2015-06-19T13:08:33+02:00)
--------------------------------------------

- Update German and Indonesian translations
- Update and improve Czech translation
- Avoid passing more arguments to g_strdup_printf() than specified in the format
- Document configuration default values
- Make the update viewer command configurable
- Prevent calling g_source_remove() on an invalid source ID
- Improve README file
- Add NEWS file

pk-update-icon 1 (2014-02-07T22:05:17+01:00)
--------------------------------------------

- Add missing function prototypes
- Stylistic fixes
- Add README file with build instructions and author information
- Convert the manpage to DocBook V5.0
- Use pax to generate tarballs
- Fix automatic dependency generation and improve portability

pk-update-icon 0.2.2 (2012-07-18T11:54:03+02:00)
------------------------------------------------

- increase the default update interval and delay before checking for updates

pk-update-icon 0.2.1 (2011-11-15T13:03:17+01:00)
------------------------------------------------

- hide the autostart file in KDE and GNOME rather than explicitly showing it in Xfce and LXDE only
- increase default startup-delay to 5 min (bnc#729230)
- fix some typos
- use the VERSION macro in the about dialog

pk-update-icon 0.2 (2011-10-22T10:20:02+02:00)
----------------------------------------------

- add dist target to Makefile
- add french translation
- added slovak and czech translations
- added a button for installing updates to the notification
- add an about dialog
- added Indonesian translations
- add CLI, make delay and interval configurable
- add clobber target
- GObject-based rewrite
- fix the locale path
- add license header to main.h
- add compatibility with libnotify version < 0.7
- add a man page
- use libunique to ensure that only a single instance of pk-update-icon can be run
- add internationalization support
- allow overriding CFLAGS
- fix desktop file and arguments order in Makefile
- use libnotify's pkgconfig for buildflags
- use GTK2 as Xfce/LXDE are not yet using GTK3
- code rework, show notify window just once
- added GPLv2 headers
- change icon and tooltip of the tray icon
- initial commit


main.c
/*
 * Copyright (C) 2011 Guido Berhoerster <gber@opensuse.org>
 *
 * Licensed under the GNU General Public License Version 2
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
 */

#include <locale.h>
#include <glib/gi18n.h>
#include <gtk/gtk.h>
#include <unique/unique.h>
#include <libnotify/notify.h>
#include "pkui-icon.h"

int
main(int argc, char **argv)
{
	PkuiIcon	*icon;
	UniqueApp	*app = NULL;
	int		exitval = 0;
	gboolean	version = FALSE;
	gchar		*update_viewer_command = NULL;
	gint		startup_delay = 300;
	gint		check_interval = 86400;
	GOptionContext	*context;
	GError		*error = NULL;
	const GOptionEntry options[] = {
		{ "command", 'c', 0, G_OPTION_ARG_STRING,
		    &update_viewer_command, N_("Command for starting the "
		    "software update viewer"), "command" },
		{ "delay", 'd', 0, G_OPTION_ARG_INT, &startup_delay,
		    N_("Set the delay in seconds before the first check for "
		    "updates"), "delay" },
		{ "interval", 'i', 0, G_OPTION_ARG_INT, &check_interval,
		    N_("Set the interval in seconds between checks for "
		    "updates"), "interval" },
		{ "version", 'v', 0, G_OPTION_ARG_NONE, &version,
		    N_("Print the version number and exit"), NULL },
		{ NULL, 0, 0, G_OPTION_ARG_NONE, NULL, NULL, 0 }
	};

	setlocale(LC_ALL, "");

	bindtextdomain(PACKAGE, LOCALEDIR);
	bind_textdomain_codeset(PACKAGE, "UTF-8");
	textdomain(PACKAGE);

	gtk_init(&argc, &argv);

	context = g_option_context_new(_("- display notifications about "
	    "software updates"));
	g_option_context_add_main_entries(context, options, PACKAGE);
	g_option_context_parse(context, &argc, &argv, &error);
	g_option_context_free(context);
	if (error) {
		g_printerr("Error parsing command line options: %s\n",
		    error->message);
		g_error_free(error);
		exitval = 1;
		goto out;
	}

	if (startup_delay < 0) {
		g_printerr("Error parsing command line options: delay must be "
		    "greater or equal to zero\n");
		exitval = 1;
		goto out;
	}

	if (check_interval < 0) {
		g_printerr("Error parsing command line options: interval must "
		    "be greater or equal to zero\n");
		exitval = 1;
		goto out;
	}

	if (version) {
		g_print("%s %s\n", PACKAGE, VERSION);
		goto out;
	}

	app = unique_app_new(APP_NAME, NULL);
	if (unique_app_is_running(app)) {
		g_printerr("Another instance of pk-update-icon is already "
		    "running, exiting\n");
		exitval = 1;
		goto out;
	}

	notify_init(PACKAGE);

	icon = pkui_icon_new(startup_delay, check_interval,
	    update_viewer_command);

	gtk_main();

	g_object_unref(icon);
out:
	if (app != NULL)
		g_object_unref(app);
	if (notify_is_initted())
		notify_uninit ();
	g_free(update_viewer_command);

	return (exitval);
}


pkui-icon.h
/*
 * (C) 2011 Guido Berhoerster <gber@opensuse.org>
 *
 * Licensed under the GNU General Public License Version 2
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
 */

#ifndef	__PKUI_ICON_H
#define	__PKUI_ICON_H

#include <glib-object.h>
#include <gtk/gtk.h>

G_BEGIN_DECLS

#define	PKUI_TYPE_ICON			(pkui_icon_get_type())
#define	PKUI_ICON(obj)			(G_TYPE_CHECK_INSTANCE_CAST((obj), \
	PKUI_TYPE_ICON, PkuiIcon))
#define	PKUI_IS_ICON(obj)		(G_TYPE_CHECK_INSTANCE_TYPE((obj), \
	PKUI_TYPE_ICON))
#define	PKUI_ICON_CLASS(klass)		(G_TYPE_CHECK_CLASS_CAST((klass), \
	PKUI_TYPE_ICON, PkuiIconClass))
#define	PKUI_IS_ICON_CLASS(klass)	(G_TYPE_CHECK_CLASS_TYPE((klass), \
	PKUI_TYPE_ICON))
#define	PKUI_ICON_GET_CLASS(obj)	(G_TYPE_INSTANCE_GET_CLASS((obj), \
	PKUI_TYPE_ICON, PkuiIconClass))

typedef struct _PkuiIcon	PkuiIcon;
typedef struct _PkuiIconClass	PkuiIconClass;
typedef struct _PkuiIconPrivate	PkuiIconPrivate;

struct _PkuiIcon
{
	GObject		parent_instance;
	PkuiIconPrivate	*priv;
};

struct _PkuiIconClass
{
	GObjectClass	parent_class;
};

GType		pkui_icon_get_type(void);
PkuiIcon	*pkui_icon_new(guint, guint, const gchar *);

G_END_DECLS

#endif /* __PKUI_ICON_H */


pkui-backend.h
/*
 * (C) 2011 Guido Berhoerster <gber@opensuse.org>
 *
 * Licensed under the GNU General Public License Version 2
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
 */

#ifndef	__PKUI_BACKEND_H
#define	__PKUI_BACKEND_H

#include <glib-object.h>
#include <packagekit-glib2/packagekit.h>

G_BEGIN_DECLS

#define	PKUI_TYPE_BACKEND		(pkui_backend_get_type())
#define	PKUI_BACKEND(obj)		(G_TYPE_CHECK_INSTANCE_CAST((obj), \
	PKUI_TYPE_BACKEND, PkuiBackend))
#define	PKUI_IS_BACKEND(obj)		(G_TYPE_CHECK_INSTANCE_TYPE((obj), \
	PKUI_TYPE_BACKEND))
#define	PKUI_BACKEND_CLASS(klass)	(G_TYPE_CHECK_CLASS_CAST((klass), \
	PKUI_TYPE_BACKEND, PkuiBackendClass))
#define	PKUI_IS_BACKEND_CLASS(klass)	(G_TYPE_CHECK_CLASS_TYPE((klass), \
	PKUI_TYPE_BACKEND))
#define	PKUI_BACKEND_GET_CLASS(obj)	(G_TYPE_INSTANCE_GET_CLASS((obj), \
	PKUI_TYPE_BACKEND, PkuiBackendClass))

typedef struct _PkuiBackend		PkuiBackend;
typedef struct _PkuiBackendClass	PkuiBackendClass;
typedef struct _PkuiBackendPrivate	PkuiBackendPrivate;

struct _PkuiBackend
{
	GObject			parent_instance;
	PkuiBackendPrivate	*priv;
};

struct _PkuiBackendClass
{
	GObjectClass	parent_class;
};

GType		pkui_backend_get_type(void);
PkuiBackend	*pkui_backend_new(guint startup_delay, guint interval);
guint		pkui_backend_get_updates_normal(PkuiBackend *self);
guint		pkui_backend_get_updates_important(PkuiBackend *self);
void		pkui_backend_set_inhibit_check(PkuiBackend *self,
    gboolean inhibit_check);
gboolean	pkui_backend_get_inhibit_check(PkuiBackend *self);
void		pkui_backend_set_check_interval(PkuiBackend *self,
    guint check_interval);
guint		pkui_backend_get_check_interval(PkuiBackend *self);
guint		pkui_backend_get_startup_interval(PkuiBackend *self);
void		pkui_backend_check_now(PkuiBackend *self);

G_END_DECLS

#endif /* __PKUI_BACKEND_H */


pkui-icon.c
/*
 * (C) 2011 Guido Berhoerster <gber@opensuse.org>
 *
 * Licensed under the GNU General Public License Version 2
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
 */

#include <glib/gi18n.h>
#include <libnotify/notify.h>
#include "pkui-backend.h"
#include "pkui-icon.h"

G_DEFINE_TYPE(PkuiIcon, pkui_icon, G_TYPE_OBJECT)

#ifndef	UPDATE_VIEWER_COMMAND
#define	UPDATE_VIEWER_COMMAND	""
#endif /* UPDATE_VIEWER_COMMAND */

#define	PKUI_ICON_GET_PRIVATE(obj)	(G_TYPE_INSTANCE_GET_PRIVATE((obj), \
					    PKUI_TYPE_ICON, PkuiIconPrivate))

struct _PkuiIconPrivate
{
	PkuiBackend	*backend;

	GtkStatusIcon	*status_icon;
	GtkWidget	*status_icon_popup_menu;
	NotifyNotification *notification;
	gchar		*update_viewer_command;
};

static GtkWidget*	icon_popup_menu_create(PkuiIcon *self);
static void	icon_popup_menu_popup(GtkStatusIcon *status_icon, guint button,
    guint activate_time, gpointer user_data);
static void	icon_activated(GtkStatusIcon *status_icon, gpointer user_data);
static void	hide_notification(PkuiIcon *self);
static void	backend_state_changed(PkuiBackend *backend, gpointer user_data);
static void	update_viewer_exited(GPid pid, gint status, gpointer user_data);

static void
pkui_icon_finalize(GObject *gobject)
{
	PkuiIcon	*self = PKUI_ICON(gobject);

	gtk_widget_destroy(self->priv->status_icon_popup_menu);
	g_object_unref(self->priv->status_icon_popup_menu);
	g_object_unref(self->priv->status_icon);
	if (self->priv->notification != NULL) {
		notify_notification_close(self->priv->notification, NULL);
		g_object_unref(self->priv->notification);
	}
	g_object_unref(self->priv->backend);
	g_free(self->priv->update_viewer_command);

	G_OBJECT_CLASS(pkui_icon_parent_class)->finalize(gobject);
}

static void
pkui_icon_class_init(PkuiIconClass *klass)
{
	GObjectClass	*gobject_class = G_OBJECT_CLASS(klass);

	gobject_class->finalize = pkui_icon_finalize;

	g_type_class_add_private(klass, sizeof (PkuiIconPrivate));
}

static void
pkui_icon_init(PkuiIcon *self)
{
	self->priv = PKUI_ICON_GET_PRIVATE(self);

	gtk_window_set_default_icon_name("system-software-update");

	self->priv->status_icon_popup_menu = icon_popup_menu_create(self);
	g_object_ref(self->priv->status_icon_popup_menu);
	g_object_ref_sink(GTK_OBJECT(self->priv->status_icon_popup_menu));

	self->priv->status_icon = gtk_status_icon_new();
	gtk_status_icon_set_title(self->priv->status_icon,
	    _("Software Updates"));
	gtk_status_icon_set_visible(self->priv->status_icon, FALSE);
	g_signal_connect(G_OBJECT(self->priv->status_icon), "activate",
	    G_CALLBACK(icon_activated), self);
	g_signal_connect(G_OBJECT(self->priv->status_icon), "popup-menu",
	    G_CALLBACK(icon_popup_menu_popup), self);

	self->priv->notification = NULL;

	self->priv->backend = NULL;

	self->priv->update_viewer_command = NULL;
}

static void
exec_update_viewer(PkuiIcon *self)
{
	gchar		**argv = NULL;
	GError		*error = NULL;
	GPid		pid;

	g_return_if_fail(PKUI_IS_BACKEND(self->priv->backend));
	g_return_if_fail(self->priv->update_viewer_command != NULL);

	g_debug("executing command %s", self->priv->update_viewer_command);

	if (!g_shell_parse_argv(self->priv->update_viewer_command, NULL,
	    &argv, &error)) {
		g_warning("Could not parse command: %s", error->message);
		g_error_free(error);
		return;
	}

	if (!g_spawn_async(NULL, argv, NULL,
	    G_SPAWN_DO_NOT_REAP_CHILD | G_SPAWN_SEARCH_PATH, NULL, NULL, &pid,
	    &error)) {
		g_warning("Could not execute command: %s", error->message);
		g_error_free(error);
		return;
	}
	g_child_watch_add(pid, (GChildWatchFunc)update_viewer_exited, self);

	pkui_backend_set_inhibit_check(self->priv->backend, TRUE);
	hide_notification(self);
}

static void
about_dialog_show(GtkMenuItem *item, gpointer user_data)
{
	static const gchar	*copyright = "Copyright \xc2\xa9 2011 Guido "
	    "Berhoerster\nCopyright \xc2\xa9 2011 Pavol Rusnak\n";
	static const gchar	*authors[3] = {
	    "Guido Berhoerster <gber@opensuse.org>",
	    "Pavol Rusnak <stick@gk2.sk>",
	    NULL
	};
	static const gchar	*documenters[2] = {
	    "Guido Berhoerster <gber@opensuse.org>",
	    NULL
	};
	static const gchar	*license =
	    "Licensed under the GNU General Public License Version 2\n\n"
	    "This program is free software; you can redistribute it and/or "
	    "modify it under the terms of the GNU General Public License as "
	    "published by the Free Software Foundation; either version 2 of "
	    "the License, or (at your option) any later version.\n\n"
	    "This program is distributed in the hope that it will be useful, "
	    "but WITHOUT ANY WARRANTY; without even the implied warranty of "
	    "MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU "
	    "General Public License for more details.\n\n"
	    "You should have received a copy of the GNU General Public License "
	    "along with this program; if not, write to the Free Software "
	    "Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, "
	    "MA 02110-1301 USA.";
	const gchar	*translators = _("translators");

	if (strcmp(translators, "translators") == 0)
	    translators = NULL;

	gtk_show_about_dialog (NULL, "version", VERSION, "copyright", copyright,
	    "authors", authors, "documenters", documenters,
	    "translator-credits", translators, "license", license,
	    "wrap-license", TRUE, NULL);
}

static void
backend_check_now(GtkMenuItem *menu_item, gpointer user_data)
{
	PkuiIcon	*self = PKUI_ICON(user_data);

	g_return_if_fail(PKUI_IS_BACKEND(self->priv->backend));

	pkui_backend_check_now(self->priv->backend);
}

static GtkWidget*
icon_popup_menu_create(PkuiIcon *self)
{
	GtkWidget	*popup_menu = gtk_menu_new();
	GtkWidget	*item;
	GtkWidget	*image;

	item = gtk_image_menu_item_new_with_mnemonic(_("_Check for Updates"));
	image = gtk_image_new_from_icon_name(GTK_STOCK_REFRESH,
	    GTK_ICON_SIZE_MENU);
	gtk_image_menu_item_set_image(GTK_IMAGE_MENU_ITEM(item), image);
	gtk_menu_shell_append(GTK_MENU_SHELL(popup_menu), item);
	g_signal_connect(G_OBJECT(item), "activate",
	    G_CALLBACK(backend_check_now), self);

	item = gtk_image_menu_item_new_with_mnemonic(_("_About"));
	image = gtk_image_new_from_icon_name(GTK_STOCK_ABOUT,
	    GTK_ICON_SIZE_MENU);
	gtk_image_menu_item_set_image(GTK_IMAGE_MENU_ITEM(item), image);
	gtk_menu_shell_append(GTK_MENU_SHELL(popup_menu), item);
	g_signal_connect(G_OBJECT(item), "activate",
	    G_CALLBACK(about_dialog_show), self);

	gtk_widget_show_all(GTK_WIDGET(popup_menu));

	return (popup_menu);
}

static void
icon_popup_menu_popup(GtkStatusIcon *status_icon, guint button,
    guint activate_time, gpointer user_data)
{
	PkuiIcon	*self = PKUI_ICON(user_data);

	gtk_menu_popup(GTK_MENU(self->priv->status_icon_popup_menu), NULL, NULL,
	    NULL, NULL, button, activate_time);
}

static void
notification_handle_action(NotifyNotification *notification, gchar *action,
    gpointer user_data)
{
	PkuiIcon	*self = PKUI_ICON(user_data);

	if (strcmp(action, "install-updates") == 0) {
		exec_update_viewer(self);
	}
}

static void
update_notification(PkuiIcon *self, guint updates_normal,
    guint updates_important)
{
	gchar	*message;
	gchar	*message_a;
	gchar	*message_b;
	gchar	*title = updates_important ?
	    /* TRANSLATORS: This is a message without number mentioned */
	    ngettext("Important Software Update", "Important Software Updates",
	    updates_important + updates_normal) :
	    /* TRANSLATORS: This is a message without number mentioned */
	    ngettext("Software Update", "Software Updates", updates_important +
	    updates_normal);
	gchar	*icon = updates_important ? "software-update-urgent" :
	    "software-update-available";

	if (updates_important > 0) {
		if (updates_normal > 0) {
			message_a = g_strdup_printf(
			/*
			 * TRANSLATORS: This sentence contains two plurals.
			 * Texts related to these plurals are mixed. That is
			 * why it is split in three parts. Fill first two parts
			 * as you need, and use them as %s in the last one to
			 * construct a sentence. Note that if the first form of
			 * plural relates only to singular form, it is never
			 * used, and dedicated shorter sentences are used. */
			    ngettext("There is %d software update available,",
			    "There are %d software updates available,",
			    updates_normal + updates_important),
			    updates_normal + updates_important);
			/*
			 * TRANSLATORS: This is the sentence part in the
			 * middle, form of which is related to the first
			 * number. If your language does not need it, simply
			 * use it as space or so. */
			message_b = ngettext("of it", "of them",
			    updates_normal + updates_important);
			message = g_strdup_printf(
			/*
			 * TRANSLATORS: This forms the sentence. If you need to
			 * swap parts, use %3$s and %1$s etc. Plurals are
			 * related to second number. */
			    ngettext("%s %d %s is important.",
			    "%s %d %s are important.", updates_important),
			    message_a, updates_important, message_b);
			g_free(message_a);
		} else {
			message = g_strdup_printf(ngettext("There is %d "
			    "important software update available.",
			    "There are %d important software updates "
			    "available.", updates_important),
			    updates_important);
		}
	} else {
		message = g_strdup_printf(ngettext("There is %d software "
		    "update available.",
		    "There are %d software updates available.",
		    updates_normal), updates_normal);
	}

	gtk_status_icon_set_tooltip(self->priv->status_icon, message);
	gtk_status_icon_set_from_icon_name(self->priv->status_icon, icon);
	gtk_status_icon_set_visible(self->priv->status_icon, TRUE);

	if (self->priv->notification == NULL) {
		self->priv->notification = notify_notification_new(title,
		    message, icon
#if (NOTIFY_VERSION_MAJOR == 0 && NOTIFY_VERSION_MINOR < 7)
		    , NULL
#endif
		    );
		if (self->priv->update_viewer_command != NULL) {
			notify_notification_add_action(self->priv->notification,
			     "install-updates", ngettext("Install Update",
			     "Install Updates",
			     updates_normal + updates_important),
			     notification_handle_action, self, NULL);
		}
	} else
		notify_notification_update(self->priv->notification, title,
		    message, icon);

	notify_notification_set_timeout(self->priv->notification,
	    NOTIFY_EXPIRES_NEVER);
	notify_notification_set_urgency(self->priv->notification,
	    updates_important ? NOTIFY_URGENCY_CRITICAL :
	    NOTIFY_URGENCY_NORMAL);
	notify_notification_show(self->priv->notification, NULL);

	g_free(message);
}

static void
hide_notification(PkuiIcon *self)
{
	gtk_status_icon_set_visible(self->priv->status_icon, FALSE);
	notify_notification_close(self->priv->notification, NULL);
}


static void
backend_state_changed(PkuiBackend *backend, gpointer user_data)
{
	PkuiIcon	*self = PKUI_ICON(user_data);
	guint		updates_normal;
	guint		updates_important;

	g_return_if_fail(PKUI_IS_BACKEND(backend));

	updates_normal = pkui_backend_get_updates_normal(backend);
	updates_important = pkui_backend_get_updates_important(backend);
	if (updates_normal > 0 || updates_important > 0)
		update_notification(self, updates_normal, updates_important);
	else if (updates_normal + updates_important == 0)
		hide_notification(self);
}

static void
update_viewer_exited(GPid pid, gint status, gpointer user_data)
{
	PkuiIcon	*self = PKUI_ICON(user_data);

	g_return_if_fail(PKUI_IS_BACKEND(self->priv->backend));

	g_spawn_close_pid(pid);

	pkui_backend_check_now(self->priv->backend);
	pkui_backend_set_inhibit_check(self->priv->backend, FALSE);
}

static void
icon_activated(GtkStatusIcon *status_icon, gpointer user_data)
{
	PkuiIcon	*self = PKUI_ICON(user_data);

	if (self->priv->update_viewer_command != NULL) {
		exec_update_viewer(self);
	}
}

PkuiIcon *
pkui_icon_new(guint startup_delay, guint check_interval,
    const gchar *update_viewer_command)
{
	PkuiIcon	*icon = g_object_new(PKUI_TYPE_ICON, NULL);

	icon->priv->update_viewer_command = g_strdup((update_viewer_command !=
	    NULL) ? update_viewer_command : UPDATE_VIEWER_COMMAND);
	icon->priv->backend = pkui_backend_new(startup_delay, check_interval);
	g_signal_connect(icon->priv->backend, "state-changed",
	    G_CALLBACK(backend_state_changed), icon);

	return (icon);
}
